pkg/interop/client.go

```diff
@@ -0,0 +1,486 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interop
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sort"
+	"time"
+
+	"github.com/spf13/afero"
+	"go.thethings.network/lorawan-stack/pkg/encoding/lorawan"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/types"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const (
+	// loRaAllianceDomain is the domain of LoRa Alliance.
+	loRaAllianceDomain = "lora-alliance.org"
+
+	// LoRaAllianceJoinEUIDomain is the LoRa Alliance domain used for JoinEUI resolution.
+	LoRaAllianceJoinEUIDomain = "joineuis." + loRaAllianceDomain
+
+	// LoRaAllianceNetIDDomain is the LoRa Alliance domain used for NetID resolution.
+	LoRaAllianceNetIDDomain = "netids." + loRaAllianceDomain
+
+	defaultHTTPSPort = 443
+)
+
+// JoinServerProtocol represents the protocol used for connection to Join-Server by interop client.
+type JoinServerProtocol uint8
+
+const (
+	LoRaWANJoinServerProtocol1_0 JoinServerProtocol = iota
+	LoRaWANJoinServerProtocol1_1
+)
+
+// BackendInterfacesVersion returns the version of LoRaWAN Backend Interfaces specification version the protocol p is compliant with.
+// BackendInterfacesVersion panics if p is not compliant with LoRaWAN Backend Interfaces specification.
+func (p JoinServerProtocol) BackendInterfacesVersion() string {
+	switch p {
+	case LoRaWANJoinServerProtocol1_0:
+		return "1.0"
+	case LoRaWANJoinServerProtocol1_1:
+		return "1.1"
+	default:
+		panic(fmt.Sprintf("Join Server protocol	`%v` is not compliant with Backend Interfaces specification", p))
+	}
+}
+
+func (p *JoinServerProtocol) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var s string
+	if err := unmarshal(&s); err != nil {
+		return err
+	}
+	switch s {
+	case "LW1.1":
```

Comment:
So this is basically "LoRaWAN Backend Interfaces 1.1"? `LW1.1` is quite confusing to me as it implies LoRaWAN 1.1

Reply:
Well, it's part of LoRaWAN, right? It just happens to be called `Backend Interfaces` spec, and I doubt there's going to be any other LoRaWAN interoperability spec, hence I chose this naming scheme.
Can you give a better suggestion? `LWBI1.1` ? To me that's even more confusing and IMO the "Backend Interfaces" part is redundant here.

Reply:
It's not redundant; LoRaWAN, LoRaWAN Regional Parameters and LoRaWAN Backend Interfaces carry their own version number. Internally, LoRaWAN is specified as L2 to avoid confusion.

In V3, we clearly separated `MAC` and `PHY`, why confuse with `LW`? Why not just `BI`?